### DISCUSSION TITLE
Event handler not called asynchrously

### DISCUSSION THREAD

**Discussion by ghost** [Textualize/textual#2566] [Q&A]
I'm using self.screen.post_message(LogMessage(message)) to log to a TextLog widget.
The logging generally works, but even though it is async, the on_log_message handler is only called with delay.

```
    async def action_reschedule_all(self):
        task = asyncio.create_task(self.action_reschedule_all_task())
```

```
    async def action_reschedule_all_task(self):
        async for result in <async generator object>:
            ...
            self.screen.post_message(LogMessage(message))
```

```
    async def on_log_message(self, message: LogMessage):
        self.log(datetime.now().isoformat())
        self.query_one("#log", TextLog).write(message.message)
```

logged timestamps are as follows:
```
2023-05-15T08:57:03.723885 - action_reschedule_all_task
2023-05-15T08:57:05.669832 - action_reschedule_all_task
2023-05-15T08:57:05.669832 - on_log_message
2023-05-15T08:57:05.671770 - on_log_message
```
as you can see action_reschedule_all_task was running async, but on_log_message was not called async.
Am I doing something wrong or is this a bug?

**Comment by willmcgugan** [comment]
You may need to clarify what you are reading from those timestamps. What would you expect to see? But more importantly: *why* would you expect to see it.

**Comment by ghost** [comment]
I would expect the first on_log_message timestamp to be directly after the first action_reschedule_all_task timestamp.
Because the LogMessage was posted using post_message right there, as the timestamp of the first action_reschedule_allt_task proofs.

**Reply by willmcgugan** [reply]
That is not necessarily going to happen. `post_message` adds it to a queue, for the widget's task to pick up. When that happens depends on a number of things, such as how busy your widget / app is, and how asyncio choses to schedule things. All `post_message` guarantees is that the message will be handled at some point in the near future.

**Comment by ghost** [comment]
I see.
Is there a recommended alternative approach for scenarios where fluid updates are required on another Screen?

**Reply by willmcgugan** [reply]
There is no other way. If you are having issues with text log updates, it is unlikely to be related to posting messages. If you share your code we might be able to offer more help.

**Reply by ghost** [reply]
Please find a minimal example below:

```
import asyncio
import time
from datetime import datetime

from textual.app import App, ComposeResult
from textual.message import Message
from textual.widgets import TextLog


class LogMessage(Message):
    def __init__(self, message) -> None:
        self.message = message
        super().__init__()


class TextLogApp(App):
    def compose(self) -> ComposeResult:
        yield TextLog(id="log")

    async def on_mount(self):
        await asyncio.create_task(self.long_running_task())

    async def long_running_task(self):
        async for _ in async_range(3):
            self.screen.post_message(LogMessage(f"{datetime.now():%H:%M:%S}"))
            time.sleep(2)

    async def on_log_message(self, message: LogMessage):
        message_with_timestamp = f"on_log_message time: {datetime.now():%H:%M:%S} - long_running_task time: {message.message}"
        self.query_one("#log", TextLog).write(message_with_timestamp)


async def async_range(count):
    for i in range(count):
        yield i
        await asyncio.sleep(0.0)


app = TextLogApp()

if __name__ == "__main__":
    app.run()
```
The output is as follows:
```
on_log_message time: 11:19:06 - long_running_task time: 11:19:00                                                                                                                                                                                                                                     
on_log_message time: 11:19:06 - long_running_task time: 11:19:02                                                                                                                                                                                                                                     
on_log_message time: 11:19:06 - long_running_task time: 11:19:04                                                                                                                                                                                                                                                                                                   
```

**Reply by willmcgugan** [reply]
You are calling `time.sleep` which is not asynchronous. It will prevent asyncio from doing anything at all during that time.

If you have a blocking long running task, you should run it in a thread. The `@work` decorator will help you there.

**Reply by ghost** [reply]
I'm new to asyncio, but as far as I understand the code I actually use is non-blocking (a request using httpx).
If I replace `time.sleep` with `await asyncio.sleep` in the minimal example above I get the same results. 
Does await cause the request to become blocking?

**Reply by ghost** [reply]
Ah made a mistake with httpx, should have used AsyncClient. Now it is working with async.

**Reply by ghost** [reply]
Thanks a lot for your support, makes it really fun working with textual!

**Reply by willmcgugan** [reply]
No problem. Glad you are enjoying it!
